File: frontend/src/lib/playbackDiagnostics/schema.js
import { PLAYBACK_DIAGNOSTICS_SCHEMA_VERSION } from "./constants";
import { sanitizeClientDiagnosticPayload } from "./privacy";

export function createDiagnosticId(prefix = "diag") {
  if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
    return `${prefix}_${crypto.randomUUID().replaceAll("-", "")}`;
  }
  const random = Math.random().toString(36).slice(2);
  return `${prefix}_${Date.now().toString(36)}${random}`;
}

export function decimalNanoseconds(value) {
  if (value == null) return null;
  if (typeof value === "bigint") return value.toString(10);
  if (typeof value === "string" && /^-?\d+$/.test(value)) return value;
  if (typeof value === "number" && Number.isSafeInteger(value)) return String(value);
  throw new TypeError("Nanosecond values must be BigInt, safe integers, or decimal strings");
}

export function captureClientClock() {
  const hasPerformance = typeof performance !== "undefined";
  return {
    client_wall_time_ms: Date.now(),
    client_monotonic_time_us: hasPerformance ? performance.now() * 1_000 : null,
    client_time_origin_ms: hasPerformance && Number.isFinite(performance.timeOrigin)
      ? performance.timeOrigin
      : null,
  };
}

export function createPlaybackDiagnosticEvent({
  eventName,
  playbackSessionId,
  eventSequence,
  sourceSequence = eventSequence,
  priority = "normal",
  severity = "info",
  observationKind = "measured_client",
  payload = {},
  clock = {},
  context = {},
  capturedClock = null,
}) {
  const now = capturedClock || captureClientClock();
  const event = {
    schema_version: PLAYBACK_DIAGNOSTICS_SCHEMA_VERSION,
    event_id: createDiagnosticId("event"),
    event_name: String(eventName || "unknown").toLowerCase(),
    event_source: "client",
    severity,
    priority,
    playback_session_id: playbackSessionId,
    playback_attempt_id: context.playback_attempt_id ?? null,
    attachment_id: context.attachment_id ?? null,
    epoch_id: context.epoch_id ?? null,
    worker_id: null,
    incident_id: context.incident_id ?? null,
    decision_id: context.decision_id ?? null,
    trace_id: context.trace_id ?? null,
    span_id: context.span_id ?? null,
    parent_span_id: context.parent_span_id ?? null,
    event_sequence: Number(eventSequence),
    source_sequence: Number(sourceSequence),
    ...now,
    client_timer_resolution_us: context.client_timer_resolution_us ?? null,
    server_wall_time_ns: null,
    server_monotonic_time_ns: null,
    server_received_wall_time_ns: null,
    server_received_monotonic_time_ns: null,
    aligned_wall_time_ns: clock.aligned_wall_time_ns
      ? decimalNanoseconds(clock.aligned_wall_time_ns)
      : null,
    clock_offset_ns: clock.clock_offset_ns != null
      ? decimalNanoseconds(clock.clock_offset_ns)
      : null,
    clock_uncertainty_ns: clock.clock_uncertainty_ns != null
      ? decimalNanoseconds(clock.clock_uncertainty_ns)
      : null,
    network_rtt_ns: clock.network_rtt_ns != null
      ? decimalNanoseconds(clock.network_rtt_ns)
      : null,
    clock_generation: Number.isInteger(Number(clock.clock_generation))
      ? Number(clock.clock_generation)
      : null,
    clock_valid: typeof clock.clock_valid === "boolean" ? clock.clock_valid : null,
    clock_invalid_reason: clock.clock_invalid_reason || null,
    playhead_ms: context.playhead_ms ?? null,
    media_element_time_ms: context.media_element_time_ms ?? null,
    duration_ms: context.duration_ms ?? null,
    platform: context.platform ?? null,
    device_class: context.device_class ?? null,
    browser_family: context.browser_family ?? null,
    browser_version: context.browser_version ?? null,
    os_family: context.os_family ?? null,
    os_version: context.os_version ?? null,
    hls_engine: context.hls_engine ?? null,
    playback_mode: context.playback_mode ?? null,
    stream_mode: context.stream_mode ?? null,
    source_kind: context.source_kind ?? null,
    observation_kind: observationKind,
    measurement_method: context.measurement_method ?? null,
    measurement_resolution: context.measurement_resolution ?? null,
    measurement_uncertainty: context.measurement_uncertainty ?? null,
    sample_window_ms: context.sample_window_ms ?? null,
    capability_available: context.capability_available ?? null,
    unavailable_reason: context.unavailable_reason ?? null,
    payload: sanitizeClientDiagnosticPayload(payload),
  };
  return event;
}

File: backend/app/services/playback_diagnostics/schema.py
from __future__ import annotations

import time
import uuid
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from .constants import EVENT_SOURCES, OBSERVATION_KINDS, SCHEMA_VERSION


DecimalTimestamp = str
EventPriority = Literal["low", "normal", "high", "critical"]
EventSeverity = Literal["debug", "info", "warning", "error", "critical"]


class DiagnosticsModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class PlaybackDiagnosticEvent(DiagnosticsModel):
    schema_version: Literal[SCHEMA_VERSION] = SCHEMA_VERSION
    event_id: str = Field(min_length=8, max_length=128)
    event_name: str = Field(min_length=1, max_length=128, pattern=r"^[a-z0-9_.:-]+$")
    event_source: str = Field(min_length=1, max_length=32)
    severity: EventSeverity = "info"
    priority: EventPriority = "normal"

    playback_session_id: str = Field(min_length=8, max_length=128)
    playback_attempt_id: str | None = Field(default=None, max_length=128)
    attachment_id: str | None = Field(default=None, max_length=128)
    epoch_id: str | None = Field(default=None, max_length=128)
    worker_id: str | None = Field(default=None, max_length=128)
    incident_id: str | None = Field(default=None, max_length=128)
    decision_id: str | None = Field(default=None, max_length=128)

    trace_id: str | None = Field(default=None, max_length=128)
    span_id: str | None = Field(default=None, max_length=128)
    parent_span_id: str | None = Field(default=None, max_length=128)

    event_sequence: int = Field(ge=1)
    source_sequence: int = Field(ge=1)

    client_wall_time_ms: float | None = None
    client_monotonic_time_us: float | None = None
    client_time_origin_ms: float | None = None
    client_timer_resolution_us: float | None = None

    server_wall_time_ns: DecimalTimestamp | None = None
    server_monotonic_time_ns: DecimalTimestamp | None = None
    server_received_wall_time_ns: DecimalTimestamp | None = None
    server_received_monotonic_time_ns: DecimalTimestamp | None = None
    aligned_wall_time_ns: DecimalTimestamp | None = None
    clock_offset_ns: DecimalTimestamp | None = None
    clock_uncertainty_ns: DecimalTimestamp | None = None
    network_rtt_ns: DecimalTimestamp | None = None

    playhead_ms: float | None = Field(default=None, ge=0)
    media_element_time_ms: float | None = Field(default=None, ge=0)
    duration_ms: float | None = Field(default=None, ge=0)

    platform: str | None = Field(default=None, max_length=64)
    device_class: str | None = Field(default=None, max_length=32)
    browser_family: str | None = Field(default=None, max_length=64)
    browser_version: str | None = Field(default=None, max_length=64)
    os_family: str | None = Field(default=None, max_length=64)
    os_version: str | None = Field(default=None, max_length=64)
    hls_engine: str | None = Field(default=None, max_length=64)
    playback_mode: str | None = Field(default=None, max_length=32)
    stream_mode: str | None = Field(default=None, max_length=32)
    source_kind: str | None = Field(default=None, max_length=32)

    observation_kind: str
    measurement_method: str | None = Field(default=None, max_length=128)
    measurement_resolution: str | None = Field(default=None, max_length=128)
    measurement_uncertainty: str | None = Field(default=None, max_length=128)
    sample_window_ms: float | None = Field(default=None, ge=0)
    capability_available: bool | None = None
    unavailable_reason: str | None = Field(default=None, max_length=256)
    payload: dict[str, Any] = Field(default_factory=dict)

    @field_validator("event_source")
    @classmethod
    def validate_event_source(cls, value: str) -> str:
        if value not in EVENT_SOURCES:
            raise ValueError("Unsupported diagnostics event source")
        return value

    @field_validator("observation_kind")
    @classmethod
    def validate_observation_kind(cls, value: str) -> str:
        if value not in OBSERVATION_KINDS:
            raise ValueError("Unsupported diagnostics observation kind")
        return value

    @field_validator(
        "server_wall_time_ns",
        "server_monotonic_time_ns",
        "server_received_wall_time_ns",
        "server_received_monotonic_time_ns",
        "aligned_wall_time_ns",
        "clock_offset_ns",
        "clock_uncertainty_ns",
        "network_rtt_ns",
    )
    @classmethod
    def validate_decimal_timestamp(cls, value: str | None) -> str | None:
        if value is not None and not value.lstrip("-").isdigit():
            raise ValueError("Nanosecond timestamps must be decimal strings")
        return value


class PlaybackDiagnosticsBootstrapRequest(DiagnosticsModel):
    playback_session_id: str = Field(min_length=8, max_length=128)
    client_instance_id: str = Field(min_length=8, max_length=128)
    platform: str = Field(default="unknown", max_length=64)
    device_class: str = Field(default="unknown", max_length=32)
    browser_family: str = Field(default="unknown", max_length=64)
    browser_version: str = Field(default="", max_length=64)
    os_family: str = Field(default="unknown", max_length=64)
    os_version: str = Field(default="", max_length=64)
    hls_engine: str = Field(default="unknown", max_length=64)
    capabilities: dict[str, bool | str | int | float | None] = Field(default_factory=dict)


class PlaybackDiagnosticsBootstrapResponse(DiagnosticsModel):
    enabled: bool
    diagnostics_session_id: str
    source_id: str
    schema_version: str
    client_spool_max_bytes: int
    batch_max_events: int
    batch_max_bytes: int
    clock_algorithm: str
    server_wall_time_ns: DecimalTimestamp
    server_monotonic_time_ns: DecimalTimestamp
    ack_watermark: int = Field(ge=0)


class PlaybackDiagnosticsBatchRequest(DiagnosticsModel):
    diagnostics_session_id: str = Field(min_length=8, max_length=128)
    source_id: str = Field(min_length=8, max_length=128)
    events: list[PlaybackDiagnosticEvent]


class PlaybackDiagnosticsBatchResponse(DiagnosticsModel):
    accepted: int
    duplicate: int
    rejected: int
    out_of_order: int
    ack_watermark: int
    capacity_state: str


class PlaybackDiagnosticsClockRequest(DiagnosticsModel):
    diagnostics_session_id: str = Field(min_length=8, max_length=128)
    source_id: str = Field(min_length=8, max_length=128)
    sample_id: str = Field(min_length=4, max_length=128)
    client_send_wall_time_ms: float
    client_send_monotonic_time_us: float


class PlaybackDiagnosticsClockResponse(DiagnosticsModel):
    sample_id: str
    client_send_wall_time_ms: float
    client_send_monotonic_time_us: float
    server_receive_wall_time_ns: DecimalTimestamp
    server_receive_monotonic_time_ns: DecimalTimestamp
    server_send_wall_time_ns: DecimalTimestamp
    server_send_monotonic_time_ns: DecimalTimestamp
    monotonic_raw_time_ns: DecimalTimestamp | None


class PlaybackDiagnosticsCloseRequest(DiagnosticsModel):
    diagnostics_session_id: str = Field(min_length=8, max_length=128)
    source_id: str = Field(min_length=8, max_length=128)
    reason: str = Field(default="client_closed", max_length=128)
    final_source_sequence: int | None = Field(default=None, ge=0)


class PlaybackDiagnosticsCloseResponse(DiagnosticsModel):
    accepted: bool
    ack_watermark: int
    finalized: bool


def build_server_event(
    *,
    event_name: str,
    playback_session_id: str,
    source_sequence: int,
    event_source: str = "server",
    observation_kind: str = "measured_server",
    priority: EventPriority = "normal",
    severity: EventSeverity = "info",
    payload: dict[str, Any] | None = None,
    **identities: Any,
) -> PlaybackDiagnosticEvent:
    wall_ns = time.time_ns()
    monotonic_ns = time.monotonic_ns()
    allowed_identities = {
        key: value
        for key, value in identities.items()
        if key
        in {
            "playback_attempt_id",
            "attachment_id",
            "epoch_id",
            "worker_id",
            "incident_id",
            "decision_id",
            "trace_id",
            "span_id",
            "parent_span_id",
            "playhead_ms",
            "media_element_time_ms",
            "duration_ms",
            "platform",
            "device_class",
            "browser_family",
            "browser_version",
            "os_family",
            "os_version",
            "hls_engine",
            "playback_mode",
            "stream_mode",
            "source_kind",
            "measurement_method",
            "measurement_resolution",
            "measurement_uncertainty",
            "sample_window_ms",
            "capability_available",
            "unavailable_reason",
        }
    }
    return PlaybackDiagnosticEvent(
        event_id=uuid.uuid4().hex,
        event_name=event_name,
        event_source=event_source,
        severity=severity,
        priority=priority,
        playback_session_id=playback_session_id,
        event_sequence=source_sequence,
        source_sequence=source_sequence,
        server_wall_time_ns=str(wall_ns),
        server_monotonic_time_ns=str(monotonic_ns),
        aligned_wall_time_ns=str(wall_ns),
        observation_kind=observation_kind,
        payload=payload or {},
        **allowed_identities,
    )
